### PR TITLE
feat: Pydantic model for the serialization schema

### PR DESCRIPTION
**Pull request by aborgna-q**
This PR ports the pydantic definition from guppy (including https://github.com/CQCL/guppylang/pull/172), and updates it to sync up with the current schema.

Changes to the schema:

- Renamed the `InputExtensions` class to `ExtensionSet`
- Added a named definition for `Const`
- Dropped `DummyOp`. It was referenced in a couple field types, but the model was never defined.

The main difference with the model in `guppy` is the use of `RootModel`-based classes instead of variables assigned to a `TypeAliasType` for defining aliases on type unions.
They both work the same, but `mypy` doesn't like using variables in types.

Additionally, I updated some READMEs and added a CI check to ensure that the json in `specification/schema/` is always up to date.

### CODE REVIEW

**Comment by aborgna-q** [comment]
Sorry for the noisy metadata changes, the CI still needed some fixes to work correctly...

File: quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+
+    c: Literal["Sum"] = Field("Sum", title="ConstTag")
+    tag: int
+    typ: Type
+    vs: list["Const"]
+
+    class Config:
+        # Need to avoid random '\n's in the pydantic description
+        json_schema_extra = {
+            "description": "A Sum variant For any Sum type where this value meets the type of the variant indicated by the tag.",
```
Comment (mark-koch) [inline quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py:141]:
```suggestion
            "description": "A Sum variant. For any Sum type where this value meets the type of the variant indicated by the tag.",
```

**Review by mark-koch — APPROVED** [review]
Nice, the Python stuff looks good to me 👍  Can't comment on the CI though :D

Now, what would be the best path for Guppy to migrate to this? Publish `quantinuum_hugr` on PyPI or pull in the dependency from this repo for now?

**Comment by aborgna-q** [comment]
We could publish a `0.0.0.a1` alpha version and use it from guppy.

I would wait for the builder API before releasing `0.1.0`.

File: quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+    For any Sum type where this value meets the type of the variant indicated by the tag
+    """
+
+    c: Literal["Sum"] = Field("Sum", title="ConstTag")
+    tag: int
+    typ: Type
+    vs: list["Const"]
+
+    class Config:
+        # Need to avoid random '\n's in the pydantic description
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py]: Yep :/

File: quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py (hunk 1 above)
Comment (aborgna-q) [reply quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py]: Done :+1:

File: quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+    inputs: TypeRow = Field(default_factory=list)
+    other_outputs: TypeRow = Field(default_factory=list)
+    sum_rows: list[TypeRow] = Field(default_factory=list)
+    extension_delta: ExtensionSet = Field(default_factory=list)
+
+    def insert_port_types(self, in_types: TypeRow, out_types: TypeRow) -> None:
+        # The types will be all None because it's not dataflow, but we only care about
+        # the number of outputs. Note that we don't make use of the HUGR feature where
+        # the variant data is appended to successor input. Thus, `predicate_variants`
+        # will only contain empty rows.
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply quantinuum-hugr-py/src/quantinuum_hugr/serialization/ops.py]: :+1:

**Pull request merged**